### PR TITLE
BaSD-MS-005: create sessions logic

### PR DESCRIPTION
**Pull request by ripellizzari**
Please checkout the logic in functions:
-Create
-Delate
-List
-Update
The update function returns only the register that we modified.

Also check if the Json file needs another change.

### CODE REVIEW

**Comment by estefrare** [comment]
@ripellizzari Please use the correct format to name the commits, using the prefix `MS-000: ` + a description. The prefix is based on the Trello ticket.

File: index.js [hunk 1]
```diff
@@ -1,9 +1,21 @@
 const express = require('express');
 const app = express();
 const PORT = 4000;
+const sessions = require('./controllers/sessions')
```
Comment (estefrare) [inline index.js:2]: put the import (`const sessions = require('./controllers/sessions')`) at the top of the file, it is a convention

File: index.js [hunk 2]
```diff
@@ -1,9 +1,21 @@
...
 
 app.get('/', (req, res) => {
     res.send('La Bestia does the best readmes ever');
 })
 
-app.listen(PORT, () => console.log(`Server runing on port ${PORT}`))
\ No newline at end of file
+//Here we have the sessions
```
Comment (estefrare) [inline index.js]: remove redundant comments

File: index.js [hunk 3]
```diff
@@ -1,9 +1,21 @@
...
+app.get('/sessions/create', sessions.createSession)
+app.get('/sessions/delete', sessions.deleteSession)
+app.get('/sessions/update', sessions.updateSession)
+
+app.listen(PORT, () => console.log(`Server runing on port ${PORT}`))
+
+
+
+
+
```
Comment (estefrare) [inline index.js]: remove the multiple break lines

File: controllers/sessions.js [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+    updateSession,
+    deleteSession,
+
+}
+
+
+
+
+
+
```
Comment (estefrare) [inline controllers/sessions.js]: remove the multiple breaks lines

File: controllers/sessions.js [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+}
+
+
+module.exports = {
+    createSession,
+    listSession,
+    updateSession,
+    deleteSession,
+
```
Comment (estefrare) [inline controllers/sessions.js]: remove brack line into the module.exports object

File: controllers/sessions.js [hunk 6]
```diff
@@ -0,0 +1,90 @@
+const fs = require('fs')
+const Sessions = require('../data/sessions.json')
+
+
+
```
Comment (estefrare) [inline controllers/sessions.js]: remote the multiples break lines

File: controllers/sessions.js [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+const listSession = (req, res) => {
+
+    res.json(Sessions);
```
Comment (estefrare) [inline controllers/sessions.js]: remember return a status code

File: controllers/sessions.js [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+            res.status(201).json(newSession)
+        }
+    })
+}
+
+
+const updateSession = (req, res) => {
+    let updatedSession
+    const updatedSessions = Sessions.map((session) => {
+        if (session.idSession == req.query.idSession) {
```
Comment (estefrare) [inline controllers/sessions.js:28]: What happens if the id does not exist on the list?  
The list should not be updated and the endpoint should return with a 404 error.

File: controllers/sessions.js [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            res.status(201).json(updatedSession)
+
+        }
+    })
+
+}
+
+
+const deleteSession = (req, res) => {
+    const filterSessions = Sessions.filter(session => session.idSession != req.query.idSession)
```
Comment (estefrare) [inline controllers/sessions.js]: the same question about if the exists of id

File: data/sessions.json [hunk 10]
```diff
@@ -0,0 +1,51 @@
+[
+    {
+        "idSession": 1,
```
Comment (estefrare) [inline data/sessions.json]: since the name of the entity is `Session` is not necessary clarify it in the `id` name, just use `id` instead of `idSession`

File: data/sessions.json [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+        "idPsicologist": 103,
+        "idPostulant": 1003,
+        "date": "Tue, 14 Dec 2021 14:30:00 GMT",
+        "state": "Pending"
+    },
+    {
+        "idSession": 5,
+        "idPsicologist": 104,
+        "idPostulant": 1004,
+        "date": "Wed, 25 Oct 2021 13:30:00 GMT",
```
Comment (estefrare) [inline data/sessions.json]: use the ISO format to date. Example: `2021-11-06T18:20:55.496Z`

File: package.json [hunk 12]
```diff
@@ -23,4 +23,4 @@
   "devDependencies": {
     "nodemon": "^2.0.14"
   }
-}
+}
```
Comment (estefrare) [inline package.json:26]: remove the new break line created

**Review by estefrare — CHANGES_REQUESTED** [review]

File: package.json (hunk 12 above)
Comment (ripellizzari) [reply package.json:26]: Here you reffer to the line 26??

Comment (estefrare) [reply package.json:26]: yes, that one

File: controllers/sessions.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+                date: req.query.date || session.date,
+                state: req.query.state || session.state
+            }
+
+            return updatedSession
+        }
+        return session
+    })
+
+    if (updatedSession === undefined) res.status(404).send('Session NOT found')
```
Comment (estefrare) [inline controllers/sessions.js]: just write `if (!updatedSession)`

File: controllers/sessions.js [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+    const deletedSession = Sessions.filter(session => session.idSession == req.query.idSession)
+
+    if (deletedSession.length === 0) res.status(404).send('Session NOT found')
+
+    fs.writeFile('./data/sessions.json', JSON.stringify(filteredSessions), {}, (error) => {
+        if (error) {
+            res.status(400).send(error)
+        } else {
+            res.status(201).json(deletedSession)
+        }
```
Comment (estefrare) [inline controllers/sessions.js:64]: if you return each `res.status(...` you don't need to use an else condition.

```
if (error) {
  return res.status(400).send(error)
}
return res.status(201).json(deletedSession)
```

File: controllers/sessions.js [hunk 15]
```diff
@@ -0,0 +1,77 @@
...
+    })
+
```
Comment (estefrare) [inline controllers/sessions.js]: remove break line

File: controllers/sessions.js [hunk 16]
```diff
@@ -0,0 +1,77 @@
+const fs = require('fs')
```
Comment (estefrare) [inline controllers/sessions.js:1]: use 2 spaces of indentation for this file

**Review by estefrare — CHANGES_REQUESTED** [review]

File: controllers/sessions.js [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+const Sessions = require('../data/sessions.json')
+
+const listSession = (req, res) => {
+    res.status(201).json(Sessions)
+}
```
Comment (estefrare) [inline controllers/sessions.js:6]: 201 is for creations

**Review by estefrare — CHANGES_REQUESTED** [review]

**Review by leimeter-joaquin — APPROVED** [review]

File: controllers/sessions.js [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+            return res.status(400).send(error)
+        }
+        return res.status(201).json(updatedSession)
+    })
+}
+
+const deleteSession = (req, res) => {
+    const filteredSessions = Sessions.filter(session => session.idSession != req.query.idSession)
+    const deletedSession = Sessions.filter(session => session.idSession == req.query.idSession)
+    if (deletedSession.length === 0) res.status(404).send('Session NOT found')
```
Comment (estefrare) [inline controllers/sessions.js]:
```
if (!deletedSession.length) {
    return res.status(404).send('Session NOT found')
}
```

File: controllers/sessions.js [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+        if (error) {
+            return res.status(400).send(error)
+        }
+        return res.status(201).json(updatedSession)
+    })
+}
+
+const deleteSession = (req, res) => {
+    const filteredSessions = Sessions.filter(session => session.idSession != req.query.idSession)
+    const deletedSession = Sessions.filter(session => session.idSession == req.query.idSession)
```
Comment (estefrare) [inline controllers/sessions.js]: you can use only one filter to do that.

```
let filteredSessions
const deletedSession = Sessions.filter((session) => {
    if (session.idSession === req.query.idSession) {
        filteredSessions = session
        return false
    }
    return true
})
```

File: controllers/sessions.js (hunk 18 above)
Comment (estefrare) [reply controllers/sessions.js]: it was not fixed

File: controllers/sessions.js [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+    if (deletedSession.length === 0) res.status(404).send('Session NOT found')
+    fs.writeFile('./data/sessions.json', JSON.stringify(filteredSessions), {}, (error) => {
+        if (error) {
+            return res.status(400).send(error)
+        }
+        return res.status(201).json(deletedSession)
```
Comment (estefrare) [inline controllers/sessions.js]: use 204 status code for deletions

File: package.json [hunk 21]
```diff
@@ -19,8 +19,5 @@
   "homepage": "https://github.com/estefrare/BaSD-MindSet#readme",
   "dependencies": {
     "express": "^4.17.1"
```
Comment (estefrare) [inline package.json]: why did you remove `nodemon`?

**Review by estefrare — CHANGES_REQUESTED** [review]

**Comment by estefrare** [comment]
@ripellizzari this branch has conflicts, please resolve them

File: package.json (hunk 21 above)
Comment (ripellizzari) [reply package.json]: I thought that was what you pointed me

File: controllers/sessions.js [hunk 22]
```diff
@@ -0,0 +1,71 @@
...
+            return updatedSession
+        }
+        return session
+    })
+    if (!updatedSession) res.status(404).send('Session NOT found')
+    fs.writeFile('./data/sessions.json', JSON.stringify(updatedSessions), {}, (error) => {
+        if (error) {
+            return res.status(400).send(error)
+        }
+        return res.status(201).json(updatedSession)
```
Comment (estefrare) [inline controllers/sessions.js]: use status code 200. 
201 is for creations

**Review by estefrare — CHANGES_REQUESTED** [review]

File: package.json (hunk 21 above)
Comment (ripellizzari) [reply package.json]: I will leave the one in the development branch

**Review by estefrare — APPROVED** [review]
